File: src/main/java/com/waridley/ttv/logger/Launcher.java
package com.waridley.ttv.logger;

import com.github.philippheuer.credentialmanager.CredentialManager;
import com.github.philippheuer.credentialmanager.CredentialManagerBuilder;
import com.github.philippheuer.credentialmanager.domain.OAuth2Credential;
import com.github.philippheuer.events4j.EventManager;
import com.github.philippheuer.events4j.domain.Event;
import com.github.twitch4j.TwitchClient;
import com.github.twitch4j.TwitchClientBuilder;
import com.github.twitch4j.chat.TwitchChat;
import com.mongodb.ConnectionString;
import com.mongodb.MongoClientSettings;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoDatabase;
import com.waridley.credentials.DesktopAuthController;
import com.waridley.credentials.NamedCredentialStorageBackend;
import com.waridley.credentials.mongo.MongoCredentialStorageBackend;
import com.waridley.ttv.RefreshingProvider;
import com.waridley.ttv.TtvStorageInterface;
import com.waridley.ttv.logger.backend.local.FileMessageLogger;
import com.waridley.ttv.logger.backend.mongo.MongoMessageLogger;
import com.waridley.ttv.mongo.MongoTtvBackend;
import org.slf4j.Logger;
import picocli.CommandLine;
import picocli.CommandLine.Option;
import picocli.CommandLine.Parameters;
import picocli.CommandLine.Command;
import reactor.core.publisher.*;
import reactor.core.scheduler.Schedulers;
import reactor.util.concurrent.WaitStrategy;

import java.io.IOException;
import java.net.URISyntaxException;
import java.util.*;

@Command(name = "ttv-logger")
public class Launcher implements Runnable {
	
	private Logger log;
	
	@Parameters(index = "0", arity="1")           private String channelName = null;
	
	@Option(names = {"-r", "--redirectUrl"})      private String redirectUrl = "http://localhost";
	@Option(names = {"-p", "--infoPath"})         private String infoPath = null;
	@Option(names = {"-m", "--connectionString"}) private String connectionString = null;
	@Option(names = {"-d", "--DB", "--dbName"})   private String dbname = "ttvlogger";
	@Option(names = {"-a", "--token"})            private String oauth2Token = null;
	@Option(names = {"-t", "--logInterval"})      private long intervalMinutes = 0L;
	@Option(names = {"-c", "--chatCollection"})   private String chatCollectionName = null;
	@Option(names = {"-f", "--chatLogFile"})      private String chatLogFilePath = null;
	@Option(names = {"--printChat"})              private boolean printChat;
	
	@Option(names = "--logLevels", split = ",")
	private Map<String, String> logLevels = new HashMap<>();
	
	@Option(names = {"-i", "--clientId"})
	private String clientId = "yod6n4not2ypylyszb7rtuqnpaq6q3"; //Waridley's TtvLogger ClientID
	
	@Option(names = {"-s", "--clientSecret"}, arity = "0..1", interactive = true)
	private String clientSecret = null;
	
	@Option(names = "-v", description = {
			"Specify multiple -v options to increase verbosity.",
			"For example, `-v -v -v` or `-vvv`"})
	private boolean[] v = new boolean[0];
	
	private RefreshingProvider idProvider;
	private MongoClient mongoClient;
	private MongoDatabase db;
	private NamedCredentialStorageBackend credBackend;
	private CredentialManager credentialManager;
	
	
	public static void main(String... args) {
		new CommandLine(new Launcher(args)).execute(args);
	}
	
	public Launcher(String... args) {
		
		Properties props = System.getProperties();
//		props.put("org.slf4j.simpleLogger.defaultLogLevel", "trace");
		props.put("org.slf4j.simpleLogger.showThreadName", "false");
		props.put("org.slf4j.simpleLogger.showLogName", "false");
		props.put("org.slf4j.simpleLogger.showShortLogName", "true");
		logLevels.put("org.slf4j.simpleLogger.log." + Launcher.class.getName(), "info");
		logLevels.put("org.slf4j.simpleLogger.log." + ChatLogger.class.getName(), "info");
		logLevels.put("org.slf4j.simpleLogger.log." + WatchtimeLogger.class.getName(), "info");
		
		for(String arg : args) {
			if(v.length > 0) {
				if(v.length == 3) {
					props.put("org.slf4j.simpleLogger.defaultLogLevel", "trace");
					logLevels.put("org.slf4j.simpleLogger.log." + ChatLogger.class.getName(), "trace");
					logLevels.put("org.slf4j.simpleLogger.log." + WatchtimeLogger.class.getName(), "trace");
				} else {
					props.put("org.slf4j.simpleLogger.log." + ChatLogger.class.getName(), "debug");
					props.put("org.slf4j.simpleLogger.log." + WatchtimeLogger.class.getName(), "debug");
					if(v.length == 2) {
						props.put("org.slf4j.simpleLogger.defaultLogLevel", "debug");
					}
				}
			} else if(arg.startsWith("--chatLogCollection=")) {
				chatCollectionName = arg.substring(20);
			} else if(arg.startsWith("--chatLogFile=")) {
				chatLogFilePath = arg.substring(14);
			}
		}
		
		for(Map.Entry<String, String> level: logLevels.entrySet()) {
			System.getProperties().put("org.slf4j.simpleLogger.log." + level.getKey(), level.getValue());
		}
		
		log = org.slf4j.LoggerFactory.getLogger(Launcher.class);
		
	}
	
	
	private synchronized void init() throws URISyntaxException, IOException {
		
		idProvider = new RefreshingProvider(clientId, clientSecret, redirectUrl);
		db = connectToDatabase();
		
		credBackend = new MongoCredentialStorageBackend(db, "credentials");
		
		DesktopAuthController authController;
		if(infoPath != null) authController = new DesktopAuthController(redirectUrl + infoPath);
		else authController = new DesktopAuthController();
		
		credentialManager = CredentialManagerBuilder.builder()
				.withAuthenticationController(authController)
				.withStorageBackend(credBackend)
				.build();
		credentialManager.registerIdentityProvider(idProvider);
		
		NamedCredentialLoader namedCredentialLoader = new NamedCredentialLoader(idProvider, this::onReceivedCredential, infoPath);
		namedCredentialLoader.startCredentialRetrieval("loggerCredential");
		
	}
	
	private synchronized void onReceivedCredential(OAuth2Credential credential) {
		Optional<OAuth2Credential> enrichedCred = idProvider.getAdditionalCredentialInformation(credential);
		if(enrichedCred.isPresent()) {
			credential = enrichedCred.get();
			log.info("Retrieved chat credential for: " + credential.getUserName());
		}
		credBackend.saveCredential("loggerCredential", credential);
		
		EventManager eventManager = new EventManager(
				Schedulers.newParallel("events4j-scheduler"),
				TopicProcessor.<Event>builder()
						.name("events4j-processor")
						.waitStrategy(WaitStrategy.sleeping())
						.bufferSize(16384)
						.build(),
				FluxSink.OverflowStrategy.BUFFER);
		
		TwitchClient twitchClient = TwitchClientBuilder.builder()
				.withEventManager(eventManager)
				.withEnableHelix(true)
				.withEnableTMI(true)
				.withEnableChat(true)
				.withChatAccount(credential)
				.withClientId(clientId)
				.withClientSecret(clientSecret)
				.withRedirectUrl(redirectUrl)
				.withCredentialManager(credentialManager)
				.build();
		twitchClient.getClientHelper().enableStreamEventListener(channelName);
		
		TtvStorageInterface ttvBackend = new MongoTtvBackend(db, twitchClient.getHelix(), credential);
		
//		TwitchChat chat = new TwitchChat(twitchClient.getEventManager(), credentialManager, credential, true, Collections.emptyList());
		TwitchChat chat = twitchClient.getChat();
		
		ChatLogger chatLogger = new ChatLogger(chat, channelName);
		
		if(chatCollectionName != null) {
			log.debug("Creating MongoDB chat logger: {}", chatCollectionName);
			chatLogger.addMessageLogger(new MongoMessageLogger(
					db,
					chatCollectionName));
		}
		if(chatLogFilePath != null) {
			try {
				log.debug("Creating FileChatLogger: {}", chatLogFilePath);
				chatLogger.addMessageLogger(new FileMessageLogger(
						chatLogFilePath));
			} catch(IOException e) {
				log.error("Failed to open chat log file:", e);
			}
		}
		if(printChat) {
			log.info("Will print chat to stdout");
			chatLogger.addMessageLogger(event -> System.out.println(String.format("%s %s... %s", "[" + event.getChannel().getName() + "]", String.format("%-25s", event.getUser().getName()).replace(' ', '.'), event.getMessage())));
		}
		
		if(intervalMinutes > 0L)
			new WatchtimeLogger(ttvBackend, channelName, intervalMinutes).start();
	}
	
	public MongoDatabase connectToDatabase() {
		ConnectionString connStr = new ConnectionString(connectionString);
		MongoClientSettings settings = MongoClientSettings.builder()
				.applyConnectionString(connStr)
				.build();
		
		mongoClient = MongoClients.create(settings);
		return mongoClient.getDatabase(dbname);
	}
	
	@Override
	public void run() {
		try {
			init();
		} catch(URISyntaxException | IOException e) {
			log.error("Failed to initialize: ", e);
		}
	}
}

File: src/main/java/com/waridley/ttv/logger/Launcher.kt
package com.waridley.ttv.logger

import com.github.ajalt.clikt.core.CliktCommand
import com.github.ajalt.clikt.parameters.options.*
import com.github.ajalt.clikt.parameters.types.long
import com.github.philippheuer.credentialmanager.CredentialManager
import com.github.philippheuer.credentialmanager.CredentialManagerBuilder
import com.github.philippheuer.credentialmanager.domain.AuthenticationController
import com.github.philippheuer.credentialmanager.domain.OAuth2Credential
import com.github.philippheuer.events4j.EventManager
import com.github.philippheuer.events4j.domain.Event
import com.github.twitch4j.TwitchClientBuilder
import com.github.twitch4j.chat.events.channel.ChannelMessageEvent
import com.mongodb.ConnectionString
import com.mongodb.MongoClientSettings
import com.mongodb.client.MongoClients
import com.mongodb.client.MongoDatabase
import com.waridley.credentials.DesktopAuthController
import com.waridley.credentials.NamedCredentialStorageBackend
import com.waridley.credentials.mongo.MongoCredentialStorageBackend
import com.waridley.ttv.RefreshingProvider
import com.waridley.ttv.TtvStorageInterface
import com.waridley.ttv.logger.ChatLogger.MessageLogger
import com.waridley.ttv.logger.NamedCredentialLoader.CredentialConsumer
import com.waridley.ttv.logger.backend.local.FileMessageLogger
import com.waridley.ttv.logger.backend.mongo.MongoMessageLogger
import com.waridley.ttv.mongo.MongoTtvBackend
import org.slf4j.LoggerFactory
import reactor.core.publisher.FluxSink
import reactor.core.publisher.TopicProcessor
import reactor.core.scheduler.Schedulers
import reactor.util.concurrent.WaitStrategy
import java.util.*

class Launcher: CliktCommand() {
	
	private val channelName
			by option("-c", "--channel", help = "The name of the channel to join.").prompt("Channel name:")
	private val clientId
			by option("-i", "--client-id", help = "Your application's client ID").prompt("Client ID:")
	private val clientSecret
			by option("-s", "--client-secret", help = "Your application's client secret.").prompt("Client secret:", hideInput = true)
	private val redirectUrl
			by option("-r", "--redirect-url", help = "The redirect URL for OAuth2 code flow. Must match the URL registered with your client ID.").default("http://localhost")
	private val infoPath
			by option("-p", "--info-path", help = "The path to the page explaining the credential code flow.")
	private val dbConnStr
			by option("-m", "--connection-string", help = "Your MongoDB connection string").prompt("MongoDB connection string:")
	private val chatCollectionName
			by option("-d", "--chat-collection", help = "The name of the MongoDB collection in which to log chat.")
	private val chatLogFilePath
			by option("-f", "--chat-file", help = "The path to a file in which to log chat.")
	private val printChat
			by option("-o", "--print-chat", help = "Print chat to stdout").flag("-n")
	private val intervalMinutes
			by option("-t", "--log-interval").long().default(0L)
	
	private lateinit var idProvider: RefreshingProvider
	private lateinit var db: MongoDatabase
	private lateinit var credBackend: NamedCredentialStorageBackend
	private lateinit var authController: AuthenticationController
	private lateinit var credentialManager: CredentialManager
	
	override fun run() {
		idProvider = RefreshingProvider(clientId, clientSecret, redirectUrl)
		db = connectToDatabase(dbConnStr)
		credBackend = MongoCredentialStorageBackend(db, "credentials")
		authController= DesktopAuthController(if(infoPath != null) redirectUrl + infoPath else null)
		credentialManager = CredentialManagerBuilder.builder()
				.withAuthenticationController(authController)
				.withStorageBackend(credBackend)
				.build()
		
		credentialManager.registerIdentityProvider(idProvider)
		
		val namedCredentialLoader = NamedCredentialLoader(idProvider, object : CredentialConsumer {
			override fun consumeCredential(credential: OAuth2Credential?) {
				onReceivedCredential(credential)
			}
		}, infoPath)
		namedCredentialLoader.startCredentialRetrieval("loggerCredential")
		
	}
	
	private fun onReceivedCredential(credential: OAuth2Credential?) {
		var cred = credential
		val enrichedCred: Optional<OAuth2Credential> = idProvider.getAdditionalCredentialInformation(cred)
		if (enrichedCred.isPresent) {
			cred = enrichedCred.get()
			log.info("Retrieved chat credential for: " + cred.userName)
		}
		credBackend.saveCredential("loggerCredential", cred)
		val eventManager = EventManager(
				Schedulers.newParallel("events4j-scheduler"),
				TopicProcessor.builder<Event>()
						.name("events4j-processor")
						.waitStrategy(WaitStrategy.sleeping())
						.bufferSize(16384)
						.build(),
				FluxSink.OverflowStrategy.BUFFER)
		val twitchClient = TwitchClientBuilder.builder()
				.withEventManager(eventManager)
				.withEnableHelix(true)
				.withEnableTMI(true)
				.withEnableChat(true)
				.withChatAccount(cred)
				.withClientId(clientId)
				.withClientSecret(clientSecret)
				.withRedirectUrl(redirectUrl)
				.withCredentialManager(credentialManager)
				.build()
		twitchClient.clientHelper.enableStreamEventListener(channelName)
		val ttvBackend: TtvStorageInterface = MongoTtvBackend(db, twitchClient.helix, cred)
		//		TwitchChat chat = new TwitchChat(twitchClient.getEventManager(), credentialManager, credential, true, Collections.emptyList());
		val chat = twitchClient.chat
		val chatLogger = ChatLogger(chat, channelName)
		if (chatCollectionName != null) {
			log.debug("Creating MongoDB chat logger: {}", chatCollectionName)
			chatLogger.addMessageLogger(MongoMessageLogger(
					db,
					chatCollectionName))
		}
		if (chatLogFilePath != null) {
			log.debug("Creating FileChatLogger: {}", chatLogFilePath)
			chatLogger.addMessageLogger(FileMessageLogger(
					chatLogFilePath!!))
		}
		if (printChat) {
			log.info("Will print chat to stdout")
			chatLogger.addMessageLogger(object : MessageLogger {
				override fun logMessage(event: ChannelMessageEvent) {
					println(String.format("%s %s... %s", "[" + event.channel.name + "]", String.format("%-25s", event.user.name).replace(' ', '.'), event.message))
				}
			})
		}
		if (intervalMinutes > 0L) WatchtimeLogger(ttvBackend, channelName, intervalMinutes).start()
	}
	
	companion object {
		private val log = LoggerFactory.getLogger(Launcher::class.java)
	}
}

fun connectToDatabase(connectionURI: String): MongoDatabase {
	val connStr = ConnectionString(connectionURI)
	val settings = MongoClientSettings.builder()
			.applyConnectionString(connStr)
			.retryWrites(true)
			.build()
	val mongoClient = MongoClients.create(settings)
	return mongoClient.getDatabase("chatgame")
}

fun main(args: Array<String>) = Launcher().main(args)